tls/s2n_security_policies.c

```diff
@@ -1478,3 +1491,22 @@ S2N_RESULT s2n_security_policy_validate_cert_signature(
 
     RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
 }
+
+S2N_RESULT s2n_security_policy_validate_certificate_chain(
+        const struct s2n_security_policy *security_policy,
+        const struct s2n_cert_chain_and_key *cert_key_pair)
+{
+    RESULT_ENSURE_REF(security_policy);
+    RESULT_ENSURE_REF(cert_key_pair);
+    RESULT_ENSURE_REF(cert_key_pair->cert_chain);
+
+    struct s2n_cert *current = cert_key_pair->cert_chain->head;
+    while (current != NULL) {
+        if (!current->info.self_signed) {
```

Comment:
I definitely want to do that for certificate key, but I'm split about the signature restriction. On one hand, the signature of a self-signed cert has no impact on the security of a chain so it feels a little silly. On the other hand, It's best to roll out the least-permissive versions of a feature first to avoid backwards-incompatible changes in the future. I think I'm leaning towards validating self-signed certs signatures?

Reply:
I've waffled on this again. Currently leaning towards "no, we shouldn't enforce certificate signatures on CA's, because there is no security value in the signature of self-signed cert"

Reply:
It's a really low-stakes decision if that helps :) My only thought was that it's one less case which is kind of clean.